src/python/pants/engine/rules.py

```diff
@@ -217,9 +217,18 @@ def rule_decorator(func, **kwargs) -> Callable:
     is_goal_cls = issubclass(return_type, Goal)
 
     # Set a default canonical name if one is not explicitly provided to the module and name of the
-    # function that implements it. This is used as the workunit name.
+    # function that implements it, plus an optional suffix. This is used as the workunit name.
+    # The suffix is a convenient way to disambiguate multiple rules registered dynamically from the
+    # same static code (by overriding the inferred param types in the @rule decorator).
```

Comment:
I still think that making the rule id generic would be cleaner, but not blocking.

Reply:
Should be easy to do later if we need to. But it really is just a string.